waku/node/peer_manager/peer_manager.nim

```diff
@@ -645,38 +646,79 @@ proc connectToRelayPeers*(pm: PeerManager) {.async.} =
   await pm.connectToNodes(outsideBackoffPeers[0..<numPeersToConnect])
 
 proc prunePeerStore*(pm: PeerManager) =
-  let numPeers = toSeq(pm.peerStore[AddressBook].book.keys).len
+  let numPeers = pm.peerStore[AddressBook].book.len
   let capacity = pm.peerStore.capacity
-  if numPeers < capacity:
+  if numPeers <= capacity:
     return
 
   debug "Peer store capacity exceeded", numPeers = numPeers, capacity = capacity
-  let peersToPrune = numPeers - capacity
-
-  # prune peers with too many failed attempts
-  var pruned = 0
-  # copy to avoid modifying the book while iterating
-  let peerKeys = toSeq(pm.peerStore[NumberFailedConnBook].book.keys)
-  for peerId in peerKeys:
-    if peersToPrune - pruned == 0:
+  let pruningCount = numPeers - capacity
+  var peersToPrune: HashSet[PeerId]
+
+  # prune failed connections
+  for peerId, count in pm.peerStore[NumberFailedConnBook].book.pairs:
+    if count < pm.maxFailedAttempts:
+      continue
+
+    if peersToPrune.len >= pruningCount:
       break
-    if pm.peerStore[NumberFailedConnBook][peerId] >= pm.maxFailedAttempts:
-      pm.peerStore.del(peerId)
-      pruned += 1
 
-  # if we still need to prune, prune peers that are not connected
+    peersToPrune.incl(peerId)
+  
   let notConnected = pm.peerStore.getNotConnectedPeers().mapIt(it.peerId)
-  for peerId in notConnected:
-    if peersToPrune - pruned == 0:
+
+  var shardlessPeers: seq[PeerId]
+  var peersByShard = initTable[uint16, seq[PeerId]]()
+
+  for peer in notConnected:
+    if not pm.peerStore[ENRBook].contains(peer):
+      shardlessPeers.add(peer)
+      continue
+
+    let record = pm.peerStore[ENRBook][peer]
+
+    let rec = record.toTyped().valueOr:
+      shardlessPeers.add(peer)
+      continue
+
+    let rs = rec.relaySharding().valueOr:
+      shardlessPeers.add(peer)
+      continue
+
+    for shard in rs.shardIds:
+      peersByShard.mgetOrPut(shard, @[peer]).add(peer)
+
+  # prune not connected peers without shard
+  for peer in shardlessPeers:
+    if peersToPrune.len >= pruningCount:
       break
-    pm.peerStore.del(peerId)
-    pruned += 1
 
-  let afterNumPeers = toSeq(pm.peerStore[AddressBook].book.keys).len
+    peersToPrune.incl(peer)
+
+  # calculate the avg peers per shard
+  var total = 0
```

Comment:
just for the fun

`let total = sum(toSeq(y.values).mapIt(it.len))`

Reply:
Thank you! I tried so many thing that didn't work or were awkward.